File: Chapter_5/src/main/scala/explode_collect.scala
// import packages

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql.Row

// defining a singleton scala object containing the main method

object explode_collect{
  def main(args:Array[String]){

    // instantiating a spark session

    val spark = SparkSession.builder.appName("explode_collect_function_prac").getOrCreate()

    // defining a schema
    val schema = StructType(Array( StructField("id", IntegerType, false),
                                    StructField("values",ArrayType(IntegerType),false)
                                    ))

    // creating a static data

    val data = Seq(Row(1,Seq(2,3,4)), Row(2,Seq(3,4,5)))

    // We currently have a local Scala collection (Seq[Row]).
    // Spark DataFrames are built on distributed data (RDD[Row]) + schema.
    // Since we defined the schema explicitly, we convert the local collection
    // into an RDD[Row] using parallelize before creating the DataFrame.'

    val data_rows = spark.sparkContext.parallelize(data)

    // passing the schema & RDD[Row] to create a dataframe

    val df = spark.createDataFrame(data_rows,schema)

    // display the schema & dataframe

    df.show()
    df.printSchema()

    // ******************* EXPLODE & COLLECT FUNCTIONS *********************************

    val exploded_df = df.select(col("id"),explode(col("values")).alias("exploded_values"))

    exploded_df.show()

    val collected_df = exploded_df.select(col("id"),(col("exploded_values") + 1).alias("added_values"))
                                  .groupBy(col("id"))
                                  .agg(collect_list("added_values").alias("collected_values"))

    collected_df.show()

    // *********************** PERFORMING SAME OPERATIONS USING UDF *****************************

    // creating a UDF to add 1 in the Array of values

    val add_one_udf = (a:Seq[Int]) => { a.map(_ +1) }

    // register the udf to spark

    spark.udf.register("addoneudf",add_one_udf)

    // using the same logic in sparksql
    // therefore registering the dataframe as temporary view first

    df.createOrReplaceTempView("udf_view")

    spark.sql("""SELECT id, addoneudf(values) AS new_values FROM udf_view""").show()

 // stop the spark session
    spark.stop()
  }
}

File: Chapter_5/pyspark/explode_collect.py
# import packages

from pyspark.sql import SparkSession
from pyspark.sql.functions import *
from pyspark.sql.types import *

# creating an entry point of the application

if __name__=="__main__":

    # instantiating a sparksession

    spark = SparkSession.builder.appName("explode_collect_example").getOrCreate()

    # defining a schema

    schema = StructType([StructField("student_id",IntegerType(),False),
                         StructField("marks_in_all_subjects",ArrayType(IntegerType(),False),False)]
                        )

    # creating a static data

    data = [ (1,[23,34,45,56]), (2,[45,32,21,34]) ]

    # forming a dataframe

    df = spark.createDataFrame(data,schema)

    # display dataframe & schema

    df.show()
    df.printSchema()

    # ************************ EXPLODE & COLLECT FUNCTIONS *********************************

    # 1. exploding values in the array

    exploded_df = df.select(col("student_id"), explode(col("marks_in_all_subjects")).alias("exploded_marks"))

    exploded_df.show()

    # 1.2 collecting the exploded array & performing mathematical operation on it

    collected_df = (exploded_df.select(col("student_id"), (col("exploded_marks") + 100).alias("add_marks")) 
                              .groupBy(col("student_id"))
                              .agg(collect_list(col("add_marks")).alias("new_marks")) 
                              .show()
                    )

    # ************************************* USER-DEFINED FUNCTION ***************************

    # 2. Acheiving the same with user-defined functions

    # creating a user-defined function

    def add_marks(a) :
        return [i + 100 for i in a]

    # register this function as UDF

    spark.udf.register("adding_marks", add_marks)

    # creating a temporary view of the dataframe

    df.createOrReplaceTempView("student_marks")

    # querying the temp view using UDF

    spark.sql(""" SELECT student_id, adding_marks(marks_in_all_subjects) FROM student_marks""").show()








    # stop the spark session
    spark.stop()


